kirkwood: add support for Seagate BlackArmor NAS220

kapieyow opened the pull request:
The Seagate BlackArmor NAS220 is a consumer NAS
with two internal drive bays. The stock OS runs
RAID 1 over the disks via mdadm.

Device specification:
- SoC: Marvell 88F6192 800 MHz
- RAM: 128 MB
- Flash: 32 MB
- 2 x internal SATA II drives
- Ethernet: 10/100/1000 Mbps (single port, no switch)
- WLAN: None
- LED: Power, Status, Sata Activity
- Key: Power, Reset
- Serial: 10 pin header, (115200,8,N,1), 3.3V TTL
	9|x  -   x|10
	7|x  -   x|8
	5|x  - GND|6
	3|x  -  RX|4
	1|TX -   x|2
	front of case
- USB ports: 2 x USB 2.0

Flash instruction:

NOTE: this process uses a serial connection. It will upgrade the
bootloader and reset the bootloader environment variables

TFTP server setup
- Setup PC with TFTP server set the PC IP to 10.4.50.5 as TFTP server
- Copy these files to TFTP server location
    - u-boot.kwb
    - seagate_blackarmor-nas220-initramfs-uImage
    - seagate_blackarmor-nas220-squashfs-sysupgrade.bin
    - seagate_blackarmor-nas220-squashfs-factory.bin

Seagate NAS setup
- Connect LAN cable between PC and seagate device
- Connect to serial to seagate device

Install u-boot
- Boot seagate device and stop in bootloader by pressing any key
- run 'printenv' from u-boot and save the values
- tftpboot 0x2000000 u-boot.kwb
- nand erase.part uboot
- nand write 0x2000000 0x0 ${filesize}
- reset

Update MAC address in u-boot env
- Stop in u-boot by pressing any key
- Get your MAC address from your saved printenv. Is also on chassis
- setenv ethaddr <your MAC>
- saveenv

Option 1 (recommended) - Install OpenWrt via initramfs and sysupgrade
- tftpboot 0x2000000 seagate_blackarmor-nas220-initramfs-uImage
- bootm 0x2000000
- *OpenWrt should be running now, however it is not written to flash yet*
- From the running instance of OpenWrt use Luci's "flash image" feature
    from the web site or use sysupgrade from the console to write
    seagate_blackarmor-nas220-squashfs-sysupgrade.bin to flash

Option 2 - Install OpenWrt by flashing factory image from u-boot
- nand erase.part ubi
- tftpboot 0x2000000 seagate_blackarmor-nas220-squashfs-factory.bin
- nand write 0x2000000 ubi ${filesize}
- reset

Signed-off-by: Kip Porterfield <kip.porterfield@gmail.com>

adrianschmutzler commented on target/linux/kirkwood/image/Makefile:
```diff
@@ -115,6 +115,15 @@ define Device/raidsonic_ib-nas62x0
+define Device/seagate_blackarmor-nas220
+  DEVICE_VENDOR := Seagate
+  DEVICE_MODEL := Blackarmor NAS220
+  BOARD_NAME := blackarmor-nas220
```
Please use DEVICE_DTS instead and drop BOARD_NAME

adrianschmutzler commented on target/linux/kirkwood/image/Makefile:
```diff
@@ -115,6 +115,15 @@ define Device/raidsonic_ib-nas62x0
+  DEVICE_MODEL := Blackarmor NAS220
+  BOARD_NAME := blackarmor-nas220
+  DEVICE_PACKAGES := kmod-hwmon-adt7475 kmod-fs-ext4 kmod-ata-marvell-sata \
+  mdadm kmod-gpio-button-hotplug
```
Please have the second line indented by a single tab.

adrianschmutzler commented on target/linux/kirkwood/patches-5.4/203-blackarmor-nas220.patch:
```diff
@@ -0,0 +1,92 @@
++			linux,default-trigger = "disk-activity";
++		};
++
++		led_healthy: healthy {
```
Please keep the label and the node name consistent, i.e. name this
led_status_blue: status_blue {

Same for the following LED.

adrianschmutzler commented on target/linux/kirkwood/patches-5.4/203-blackarmor-nas220.patch:
```diff
@@ -0,0 +1,92 @@
+ 	status = "okay";
++
++	partitions {
++		compatible = "fixed-partitions";
```
I'd expect an empty line between compatible and first partiton node.

adrianschmutzler commented on target/linux/kirkwood/patches-5.4/203-blackarmor-nas220.patch:
```diff
@@ -0,0 +1,92 @@
++			read-only;
++		};
++
++		partition@0xa0000 {
```
I don't think it's common to include the 0x here, just use partition@a0000 ...

adrianschmutzler commented on target/linux/kirkwood/patches-5.4/203-blackarmor-nas220.patch:
```diff
@@ -0,0 +1,92 @@
++			read-only;
++		};
++
++		partition@0xc0000 {
```
What with 0xb0000 to 0xc0000?

adrianschmutzler commented:
Hmm, do you know why support for this device in kernel is incomplete? Can you please upstream the changes you propose here as well.

adrianschmutzler commented:
Please use the proper name for the commit title as well:
"kirkwood: add support for Seagate BlackArmor NAS220"

kapieyow commented:
Thanks for the fast review. I'll get these changes in for re-review, but might be a couple days. I had the "NAS" in the commit title initially but was trying to get it <=50 chars. I attempt to answer your questions below. Please let me know if you have other questions and I can try and provide more. 

- The odd gap from 0xb0000 to 0xc0000 is what the stock firmware does. You can also see it in the upstream u-boot source. I don't know its purpose but I left it thinking it may provide benefits for reversion to stock firmware. 

- I don't know why the upstream does not have nand partitions, the upstream u-boot does specify them. There were some other minor things like the linux button codes didn't make sense. I have not tried to run the upstream with little/no change but could if that makes sense. 

- For background on the partitions. The stock partitions have preroot, uimage and rootfs (see below). I repartitioned in u-boot and openwrt to make one ubi partition instead of separate uimage/rootfs. My intent was to enable UBIFS over nand and allow openwrt to flex the kernel size etc. I am certainly open to doing it differently if there is a more appropriate method. 

![image](https://user-images.githubusercontent.com/33786261/86484877-38a50580-bd25-11ea-9b14-d5c9427c2f64.png)

kapieyow commented:
One more note, in case it helps. I attempted to use the seagate goflexnet and dockstart as guides. They are seagate kirkwood NAS-ish devices that are openwrt supported.  They patch u-boot and go to ubi partitions also, however it looks like upstream has partitions for these.

adrianschmutzler commented:
> The odd gap from 0xb0000 to 0xc0000 is what the stock firmware does. You can also see it in the upstream u-boot source. I don't know its purpose but I left it thinking it may provide benefits for reversion to stock firmware.

Then maybe just add a "reserved" partition with read-only for that area.

kapieyow commented:
@adrianschmutzler I believe I applied all the changes you were looking for. Please review and let me know if I have missed anything. Thanks.

wulfy23 commented:
nice work! minor observation... last three lines of the commit message ( flash instructions ) can be clearer... people should boot from initramfs then flash xyz-factory.bin ?

kapieyow commented:
Good point. Thanks for the feedback. In the current instructions I am tftp-ing the initramfs and then flashing the image, which isn't entirely necessary. Let me know if the revised steps below are more clear. The main change is to directly write the factory image to nand from u-boot, which skips the initramfs step and hopefully simplifies the process. I will do a run later today to verify my addressing etc. If it looks better to you, I will update the commit message. I am going to add more detail on the wiki too if this PR goes through ok.


```
Flash instruction:

NOTE: this process uses a serial connection. It will upgrade the
bootloader and reset the bootloader environment variables

TFTP server setup
- Setup PC with TFTP server set the PC IP to 10.4.50.5 as TFTP server
- Copy u-boot.kwb and openwrt-kirkwood-seagate_blackarmor-nas220-squashfs-factory.bin to TFTP server location

Seagate NAS setup
- Connect LAN cable between PC and seagate device
- Connect to serial to seagate device

Install u-boot
- Boot seagate device and stop in bootloader by pressing any key
- run 'printenv' from u-boot and save the existing values
- tftpboot 0x2000000 u-boot.kwb
- nand erase 0x0 0xa0000
- nand write 0x2000000 0x0 0xa0000
- reset
	- This will overwrite the u-boot env based on bad CRC

Install OpenWrt
- Stop in u-boot by pressing any key
- setenv ethaddr - also in orig printenv
- saveenv
- tftpboot 0x2000000 openwrt-kirkwood-seagate_blackarmor-nas220-squashfs-factory.bin
- nand erase.part ubi
- nand write 0x2000000 ubi 0x1e80000
- reset
```

wulfy23 commented:
Typically, the tftp from initramfs... then boot(cmd)... then perform the flash from the live system is more foolproof / common ( like you had initially )... that said... these instructions are also handy... especially when there is no wiki... 

Was mostly just thinking the wording you had/have was a little ambiguous... as to whether we are in a booted system or not... and whether we are flashing a factory/sysupgrade image...

Flash openwrt
- In running openwrt sysupgrade from CLI or Luci
openwrt-kirkwood-seagate_blackarmor-nas220-squashfs-factory.bin

For example might be re-worded to;

-Boot into OpenWrt initramfs from uboot ( with XYZ command )
-Use the FACTORY?(are you sure you mean factory here... from a booted initramfs... i would have guessed the non-factory would be optimal) image ( from LUCI or CLI ) to install permanently onto flash

bobafetthotmail commented:
>Flash openwrt
>- In running openwrt sysupgrade from CLI or Luci
>openwrt-kirkwood-seagate_blackarmor-nas220-squashfs-factory.bin

Pretty sure you cannot sysupgrade with a factory bin image so I think this is a typo. Facotry image is supposed to be flashed raw and lacks the metadata needed for a sysupgrade.

kapieyow commented:
Thanks all. You are completely correct. The flash instructions using the **initramfs** in the commit message should 
use the **sysupgrade** image. I mistakenly put **factory** in there. My apologies. The **initramfs** process using **sysupgrade** image works correctly,, however through this I realized the **factory** image will not work. I missed that the NAND UBI block size settings on this device are not the kirkwood target defaults and there are some
other UBI image building aspects I need to correct. 

I plan to correct the make image building for the factory image so that both the initramfs (already works)
and the direct flash of factory options work completely and then update the flash instructions to offer both options and
using the correct images in each case. This will require a re-push. 

@adrianschmutzler I plan to correct and force push to this PR, but if you prefer to close this PR and I submit a new one please let me know

adrianschmutzler commented:
> I plan to correct and force push to this PR, but if you prefer to close this PR and I submit a new one please let me know

Please use the existing PR as this will keep the discussion history to the new state.

kapieyow commented:
Oops. Merge conflict. Working on a fetch and re-push

kapieyow commented:
Ok. This is ready now for re-review. Thanks to all for the questions and tips. I did have the factory image setup incorrectly. I sometimes forget how much magic sysupgrade does. I made changes for factory image and ran through install instructions for it and the other sysupgrade a few times now. If you need more info let me know.

adrianschmutzler commented on target/linux/kirkwood/image/Makefile:
```diff
@@ -126,6 +126,19 @@ define Device/raidsonic_ib-nas62x0
+define Device/seagate_blackarmor-nas220
+  DEVICE_VENDOR := Seagate
+  DEVICE_MODEL := Blackarmor NAS220
+  DEVICE_DTS := kirkwood-blackarmor-nas220
```
I've just pushed a patch to "calculate" DEVICE_DTS automatically from the device node name, so this line can be dropped.

kapieyow replied:
Removed the DEVICE_DTS line. Let me know if you need anything else. Thanks

kapieyow committed:
kirkwood: add support for Seagate BlackArmor NAS220

The Seagate BlackArmor NAS220 is a consumer NAS
with two internal drive bays. The stock OS runs
RAID 1 over the disks via mdadm.

Device specification:
- SoC: Marvell 88F6192 800 MHz
- RAM: 128 MB
- Flash: 32 MB
- 2 x internal SATA II drives
- Ethernet: 10/100/1000 Mbps (single port, no switch)
- WLAN: None
- LED: Power, Status, Sata Activity
- Key: Power, Reset
- Serial: 10 pin header, (115200,8,N,1), 3.3V TTL
	9|x  -   x|10
	7|x  -   x|8
	5|x  - GND|6
	3|x  -  RX|4
	1|TX -   x|2
	front of case
- USB ports: 2 x USB 2.0

Flash instruction:

NOTE: this process uses a serial connection. It will upgrade the
bootloader and reset the bootloader environment variables

TFTP server setup
- Setup PC with TFTP server set the PC IP to 10.4.50.5 as TFTP server
- Copy these files to TFTP server location
    - u-boot.kwb
    - seagate_blackarmor-nas220-initramfs-uImage
    - seagate_blackarmor-nas220-squashfs-sysupgrade.bin
    - seagate_blackarmor-nas220-squashfs-factory.bin

Seagate NAS setup
- Connect LAN cable between PC and seagate device
- Connect to serial to seagate device

Install u-boot
- Boot seagate device and stop in bootloader by pressing any key
- run 'printenv' from u-boot and save the values
- tftpboot 0x2000000 u-boot.kwb
- nand erase.part uboot
- nand write 0x2000000 0x0 ${filesize}
- reset

Update MAC address in u-boot env
- Stop in u-boot by pressing any key
- Get your MAC address from your saved printenv. Is also on chassis
- setenv ethaddr <your MAC>
- saveenv

Option 1 (recommended) - Install OpenWrt via initramfs and sysupgrade
- tftpboot 0x2000000 seagate_blackarmor-nas220-initramfs-uImage
- bootm 0x2000000
- *OpenWrt should be running now, however it is not written to flash yet*
- From the running instance of OpenWrt use Luci's "flash image" feature
    from the web site or use sysupgrade from the console to write
    seagate_blackarmor-nas220-squashfs-sysupgrade.bin to flash

Option 2 - Install OpenWrt by flashing factory image from u-boot
- nand erase.part ubi
- tftpboot 0x2000000 seagate_blackarmor-nas220-squashfs-factory.bin
- nand write 0x2000000 ubi ${filesize}
- reset

Signed-off-by: Kip Porterfield <kip.porterfield@gmail.com>

kapieyow commented:
@adschm and @ynezz do you need other changes for this PR?

wulfy23 commented:
LGTM pending maintenance re:upstream differentials... which seem pretty unlikely/infrequent/and trivial if they arise...

kapieyow commented:
Oops I missed this one earlier. It has 2 internal SATA II bays. Stock uses mdadm raid 1 by default and you can remount them as they are if you want on openwrt. There are some small partitions the orig OS used that won't be used in that case. Think it is worth listing that in the commit message?

wulfy23 commented:
'2 x SATAII internal' should be listed in specs...
comment re: 'oem: mdadm raid 1' also useful...

otherwise other dev's if touching this device later won't be able to ascertain why these inclusions are present... set's a good example for other NAS additions also... as people may be searching for 'dual drive' NAS etc. etc.

kapieyow commented:
Excellent points @wulfy23. Commit message and first PR comment updated.

ynezz commented:
Thanks! Pulled into my staging tree at https://git.openwrt.org/openwrt/staging/ynezz.git
